include/oneapi/dpl/internal/distributed_ranges_impl/sp/algorithms/for_each.hpp

```diff
@@ -0,0 +1,94 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/sycl.hpp>
+
+#include "../../detail/sycl_utils.hpp"
+#include "../detail.hpp"
+#include "../init.hpp"
+#include "../util.hpp"
+#include "../zip_view.hpp"
+#include "execution_policy.hpp"
+
+namespace oneapi::dpl::experimental::dr::sp
+{
+
+template <typename ExecutionPolicy, distributed_range R, typename Fn>
+void
+for_each(ExecutionPolicy&& policy, R&& r, Fn fn)
+{
+    static_assert( // currently only one policy supported
+        std::is_same_v<std::remove_cvref_t<ExecutionPolicy>, sycl_device_collection>);
+
+    std::vector<sycl::event> events;
+
+    for (auto&& segment : ranges::segments(r))
+    {
+        auto&& q = __detail::queue(ranges::rank(segment));
+
+        assert(stdrng::distance(segment) > 0);
+
+        auto local_segment = __detail::local(segment);
+
+        auto first = stdrng::begin(local_segment);
+
+        auto event = dr::__detail::parallel_for(q, sycl::range<>(stdrng::distance(local_segment)),
+                                                [=](auto idx) { fn(*(first + idx)); });
+        events.emplace_back(event);
```

Comment:
Please, see my comment above (about reservation capacity for std::vector)

Reply:
applied